### ISSUE TITLE
Document: Jakarta EE 10: Support Jakarta Faces 4.0 (10/21/22 asked for info)

### ISSUE THREAD

**Issue by pnicolucci** [OpenLiberty/docs#5837]
**Please describe the problem you are having with the documentation. Is information missing, inaccurate, or unclear? Tell us about the context where you encountered the problem so we can understand how to address it.**

https://github.com/OpenLiberty/open-liberty/issues/18731

**Comment by chirp1** [comment]
Slacked Paul, included David, asking when Paul cn provide the info to doc.

**Comment by volosied** [comment]
FYI: TCK work is about wrapped up, so I'll start on this doc issue now.

**Comment by volosied** [comment]
`**~~~~~~~~~~~~ DRAFT! ~~~~~~~~~~~**`
#  Faces 4.0 Changes

## New 
### Features
  - Programmatically create facelets. 
    -  Facelets no longer need to exist as xml files and can now be programmatically created via Java. Please see the spec issue for an example: `https://github.com/jakartaee/faces/issues/1581`. Note: MyFaces escape special characters (such as `<` and `>`) by default, so the developer will need to set the escape attribute as false. This oversight by the community will be followed up. 
  - Automatic extensionless mapping
    - Facelets can now be reached without any extensions. It is configured by setting the `jakarta.faces.AUTOMATIC_EXTENSIONLESS_MAPPINGS` context parameter to true. For instance, somePage.xhtml can be reached simply via `<context-root>/somePage` since the FacesServlet will map to `/somePage`  automatically. 
### Attributes
  - <h:inputFile multiple="…">
    - Multiple files can now be uploaded via the new [multiple](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/multiple) attribute. 
  - <h:inputFile accept="…">
    - Developers can now specify the comma-separated string of mime file types that the client file browser will filter through. 
  - <h:inputText type="…">
    -  This new attribute will allow [HTML5  input types](https://developer.mozilla.org/en-US/docs/Learn/Forms/HTML5_input_types), such as `email`, `date`, `tel`, and many others. 
  - <f:websocket onerror="…">
    - It was reported that the `onclose` attribute was sometimes not enough for reporting errors, so the `onerror` attribute was added to allow developers to handle any unexpected errors that occur to the websocket connection. 
### Tags
  - [<f:selectItemGroups>](https://jakarta.ee/specifications/faces/4.0/vdldoc/f/selectItemGroups.html)
    - This new tag renders multiple`<optgroup>` elements, allowing developers to offer a nested collection of `<option>` elements. [More about optgroup element can be found here. ](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/optgroup)
  - [<f:selectItemGroup>](https://jakarta.ee/specifications/faces/4.0/vdldoc/f/selectItemGroup.html)
    - This is similar to the tag above but only allows one optgroup to be rendered.  
### Methods
  - FacesContext#getLifecycle()
    - A convenience method to determine the current phase of the Faces lifecycle. 
  - UIViewRoot#getDoctype()
    - This method will return the declared DOCTYPE of the facelet. 
### Other
  - The @ClientWindowScoped annotation has been added.
    - A new CDI scope that allows classes with this annotation to live for the current ClientWindow (enabled via setting `javax.faces.CLIENT_WINDOW_MODE` to `url` mode. In other words, when the jfwid request parameter is first used and reused across faces link components.   It is similar to the FlowScope annotation but more flexible (as it's not limited to a subset of pages). It can be considered similar to CDI's ConversationScoped annotation (but without a need to end a conversation explicitly).  Faces limits the number of Client Windows via `jakarta.faces.NUMBER_OF_CLIENT_WINDOWS` (default is 10).
  - Annotation literals for all `@Qualifiers` have been added. 
    - With this change, developers no longer need to create their own Annotation Literals. Now developers can easily retrieve Faces-specific instances via the CDI API, as so: `Map<String, Object> requestMap = CDI.current().select(new TypeLiteral<Map<String, Object>>() {}, RequestMap.Literal.INSTANCE).get();` 
  - The list layout  (i.e`layout="list"`) was added to <h:selectManyCheckbox> and <h:selectOneRadio> 
    - This new layout will allow easier CSS control since the HTML is generated with `<ul>` and `<li>` elements rather than the outdated `<table>` and associate elements. 

## Improvements 
  - UIComponent#subscribeToEvent() is now more convenient. The spec now mandates that listeners can no longer be added twice, and `UIComponent#getListenersForEventClass()` must not return null. 
  - The `type` attribute is skipped for <link> and <script> elements when the doctype is HTML5, as they are already implied.  
  - The <f:ajax> behavior has been improved for composite components when used with the `@this` keyword. The following was added to the specification: `When nested within a composite component, and the value contains or implies the keyword @this, then the keyword must be remapped to the targets attribute of the associated <composite:clientBehavior> declaration...` ([source - execute attribute](https://jakarta.ee/specifications/faces/4.0/vdldoc/f/ajax.html)) See the spec issue for additional details: `https://github.com/jakartaee/faces/issues/1567`. 
  - ExternalContext#addResponseCookie() supports custom cookie attributes, such as SameSite. This is accomplished via Servlet 6.0's [jakarta.servlet.http.Cookie#setAttribute()](https://jakarta.ee/specifications/servlet/6.0/apidocs/jakarta.servlet/jakarta/servlet/http/cookie#setAttribute(java.lang.String,java.lang. String).
  
## Deprecations
 - No new deprecation, but many removals have occurred for existing deprecated methods and classes. 

## Removals 
 - _Note: Links reference Faces 3.0 API Docs for convenience as these are removed from Faces 4.0._
 - Pages (formerly JSP) support has been removed entirely. Pages (JSP) can no longer be used as a view technology. Only facelets are now supported. 
 - Native Managed Bean support (and related code) has been removed. Developers must now switch to CDI.
 - MethodBinding, ValueBinding, and related fields/classes have been removed. Developers should switch the Expression Language equivalents (such as MethodBind to MethodExpression, PropertyResolver to ELResolver, and so forth). See the [Faces 3.0 API](https://jakarta.ee/specifications/faces/3.0/apidocs/jakarta/faces/el/package-summary.html) for migration assistance. 
 - The `CURRENT_COMPONENT` and `CURRENT_COMPOSITE_COMPONENT` constants have been removed from the UIComponent class. The  UIComponent#getCurrentComponent() and UIComponent#getCurrentCompositeComponent() utility methods should be used instead.
 - The deprecated StateManger methods have been removed. See the [Faces 3.0 API notes for the StateManger](https://jakarta.ee/specifications/faces/3.0/apidocs/jakarta/faces/application/statemanager) for the migration assistance.
 - The ResourceResolver class has been removed. The ResourceHandler should be used instead, as it can perform the same functionality. Once again, please see the [Faces 3.0 API docs for ResourceResolver](https://jakarta.ee/specifications/faces/3.0/apidocs/jakarta/faces/view/facelets/resourceresolver) for migration assistance. 
 - The PreJsf2ExceptionHandlerFactory class was also removed. 

## Other
 -  The term "JSF" has been changed to "Faces" throughout the code.  
    - This includes renames such as `jsf.js` to `faces.js`, `window.jsf` to `window.faces`, `xmlns:jsf` to `xmlns:faces`, and so forth. The spec issue can be found here: `https://github.com/jakartaee/faces/issues/1552`.
 -  The `http://xmlns.jcp.org/jsf/`URIs have been updated to the `jakarta.faces.` URN. However, Faces 4.0 is still backward compatible with the older URIs. 
      -   `http://xmlns.jcp.org/jsf`  - >` jakarta.faces`
      -   `http://xmlns.jcp.org/jsf/facelets`  -> `jakarta.faces.facelets`
      -   `http://xmlns.jcp.org/jsf/core`  ->  `jakarta.faces.core`
      -   `http://xmlns.jcp.org/jsf/html`  -> `jakarta.faces.html`
      -   `http://xmlns.jcp.org/jsf/passthrough`  -> `jakarta.faces.passthrough`
      -   `http://xmlns.jcp.org/jsf/composite`  -> `jakarta.faces.composite`
      -   `http://xmlns.jcp.org/jsf/component`  -> `jakarta.faces.component`
      -   `http://xmlns.jcp.org/jsp/jstl/core`  -> `jakarta.tags.core`
      -   `http://xmlns.jcp.org/jsp/jstl/functions` -> `jakarta.tags.functions`

 -  MyFaces 4.0 was extensive work done to comply with the new Faces TCK (which ported numerous tests from the Mojarra Project.) 
     - Changes include bug fixes, improvements, and (small) new features. For example, 
       - This section needs some work & more description. 
       - `https://github.com/jakartaee/faces/issues/329`
       - `https://github.com/jakartaee/faces/issues/1135` 
       - `https://github.com/eclipse-ee4j/mojarra/issues/4734`
       - `https://github.com/jakartaee/faces/issues/1422` 
       - `https://github.com/eclipse-ee4j/mojarra/issues/4330`
     

## Configuration
The following context parameters have been added: 
-  jakarta.faces.AUTOMATIC_EXTENSIONLESS_MAPPINGS
   - Enables automatic extensionless mapping. The default value is false.  
-  jakarta.faces.NUMBER_OF_CLIENT_WINDOWS
   -  Sets the max number of ClientWindows (default is 10), which is used by the ClientWindowScoped annotation. It can only be used when jakarta.faces.CLIENT_WINDOW_MODE is enabled.
- org.apache.myfaces.USE_LAMBDA_METAFACTORY 
  - Although not spec defined, this is an experimental MyFaces parameter that allows lambdas to be used for the setting and getting of bean properties rather than reflection. Lambdas offer significant performance improvement over reflection. If any problems arise, an error is logged, and a fallback (the traditional lookup) is used in order to prevent the application from breaking.
  
  ______________
  Further Resources
  - https://myfaces.apache.org/#/core40
  - https://jakarta.ee/specifications/faces/4.0/

**Comment by dmuelle** [comment]
Hi @volosied - these updates are now ready to review on the draft site:

https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/jakarta-ee10-diff.html#faces

I made some edits for style and concision and reorganized things a bit to fit with the rest of the content on that page.

I also haven't added the MyFaces TCK PRs, partially because you had a note that the section needed more work but also because they seemed a bit out of place with the content from other specs on this page. If this information has a direct impact on migrating apps to Faces 4.0, we can include it and add context. If it's more general "release notes" type information, we might be able to leave it off here. 

When you have a chance, let me know what you think and what changes/additions are needed. When you're satisfied with the section you can add the `Technical reviewed` label to this issue to sign off. Thanks!

**Comment by volosied** [comment]
@dmuelle  Looks good overall, just a few comments: 

Programmatically create facelets 
**Remove**: Note that MyFaces implementations escape special characters, such as < and > by default. Therefore, you must set the escape attribute to false to use these characters with programmatically created instances. This oversight by the community will be addressed in a future release.

**Question**: New tags have links in the titles, should we update the other section titles? 

@ClientWindowScoped
**Remove**: It is similar to the CDI @ConversationScoped annotation, but without a need to end a conversation explicitly. 

Other improvements and updates 
**Spelling**: see the _exectue_ attribute for the ajax tag  

Paul said it's fine to leave out the TCK PRs. Maybe add a sentence: "For all other non-spec changes, please see the MyFaces release notes." 

Thanks!

**Comment by dmuelle** [comment]
Hi @volosied 

- Removed the two cited passages
- added links to the other methods and parameters in titles- could you verify the links I found. Not 100% sure I ve got them right. ALso couldn't find javadoc for `org.apache.myfaces.USE_LAMBDA_METAFACTORY`
- is it `jakarta.faces.AUTOMATIC_EXTENSIONLESS_MAPPINGS` or `jakarta.faces.AUTOMATIC_EXTENSIONLESS_MAPPING` ?
- fixed "execute" typo 
- I added a link to the MyFaces rlease notes but wasn't sure which 4.0 RC to link to so I linked to the list of releases that includes links to release notes for each release. Let me know if theres a better link to use here.

Let me know if any further changes are needed. If not, you can add the Technical reviewed label to this issue to sign off. Thanks!

**Comment by pnicolucci** [comment]
@dmuelle it is `jakarta.faces.AUTOMATIC_EXTENSIONLESS_MAPPING` : https://jakarta.ee/specifications/faces/4.0/apidocs/constant-values.html#jakarta.faces.webapp.FacesServlet.AUTOMATIC_EXTENSIONLESS_MAPPING_PARAM_NAME

So we need to be consistent here and update the section `Map to facelets without specifying an extension` which current lists `jakarta.faces.AUTOMATIC_EXTENSIONLESS_MAPPINGS`

**Comment by pnicolucci** [comment]
@dmuelle `org.apache.myfaces.USE_LAMBDA_METAFACTORY` isn't spec defined,  the only documentation that is available is out on the myfaces page:  https://myfaces.apache.org/#/core40

**Comment by pnicolucci** [comment]
> I added a link to the MyFaces rlease notes but wasn't sure which 4.0 RC to link to so I linked to the list of releases that includes links to release notes for each release. Let me know if theres a better link to use here.

This looks fine to me. Gives someone the ability to see what's changed release to release.

**Comment by pnicolucci** [comment]
> added links to the other methods and parameters in titles- could you verify the links I found. Not 100% sure I ve got them right. ALso couldn't find javadoc for org.apache.myfaces.USE_LAMBDA_METAFACTORY

Other links also look good.

**Comment by volosied** [comment]
@dmuelle 

We think it's worth noting the update below in  the doc issue. 

Could you add a section under `Other improvements and updates`, perhaps? 

`
MyFaces updated it's JavaScript implementation to the newer ECMAScript 2015 (ES6) standard.  This means older browsers, such a Internet Explorer, will not support the newer syntax.  Automation tools for testing, such as HTMLUnit, may also encounter errors. 
`

**Comment by dmuelle** [comment]
Hi @ramkumar-k-9286 this ready for peer review

https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/jakarta-ee10-diff.html#faces

**Comment by ramkumar-k-9286** [comment]
Peer Review
---

1. The [Jakarta Faces 4.0](https://docs-draft-openlibertyio.mqj6zf7jocq.us-south.codeengine.appdomain.cloud/docs/latest/reference/feature/faces-4.0.html) feature introduces programmatic creation of Facelets... The link is not working ( believe it will be fixed when associated pages are introduced).

2. This annotation is similar to the `FlowScope` annotation but more flexible, because it’s not limited to a subset of pages. Faces limits the number of Client Windows with the `jakarta.faces.NUMBER_OF_CLIENT_WINDOWS` parameter, which has a default value of 10.
->
This annotation is similar to the `FlowScope` annotation but more flexible because **it is** not limited to a subset of pages. Faces **limit** the number of Client Windows with the `jakarta.faces.NUMBER_OF_CLIENT_WINDOWS` parameter, which has a default value of 10.

3. MyFaces updated its JavaScript implementation to the newer ECMAScript 2015 (ES6) standard. Be aware that older browsers, such a Internet Explorer, do not support the newer syntax. Automation tools for testing, such as HTMLUnit, also might encounter errors.
->
MyFaces updated its JavaScript implementation to the newer ECMAScript 2015 (ES6) standard. Be aware that older browsers, such **as** Internet Explorer, do not support the newer syntax. Automation tools for testing, such as HTMLUnit, also might encounter errors.

4. For all other non-specification changes, see the [MyFaces release notes](https://myfaces.apache.org/#/news).
->
For all other **nonspecification** changes, see the [MyFaces release notes](https://myfaces.apache.org/#/news). (Acrolinx suggestion).

**Comment by dmuelle** [comment]
@ramkumar-k-9286 thanks for reviewing

1. links will work after autogen update
2. contractions are ok
3. fixed
4. I think this is a false positive in Acrolinx as nonspecification is not really a word.
